infra/modules/s3/pdf_processing_lambda.zip

```diff
@@ -0,0 +1,166 @@
+const AWS = require("aws-sdk");
+const fetch = require("node-fetch");
+
+AWS.config.update({ region: process.env.AWS_REGION });
+
+exports.handler = async (event) => {
+  try {
+    const bucket = event.Records[0].s3.bucket.name;
+    const key = event.Records[0].s3.object.key;
+
+    const pdfUrl = `https://${bucket}.s3.amazonaws.com/${key}`;
+
+    const publicationId = key.replace(/\.pdf$/, "");
+
+    const publication = await fetch(
+      `https://${process.env.ENVIRONMENT}.api.octopus.ac/v1/publications/${publicationId}`,
+      { method: "GET" }
+    );
+
+    const pdfMetadata = mapPublicationToMetadata(publication, pdfUrl);
+
+    const apiResponse = await postToPubrouter(pdfMetadata);
+
+    // Check the API response and handle failures
+    if (apiResponse.status !== "success") {
+      await retryPostingPublication(pdfMetadata);
+    }
+  } catch (error) {
+    await sendFailureEmail(error, event, publicationId);
+  }
+};
+
+const postToPubrouter = async (pdfMetadata) => {
+  const apiEndpoint = `https://pubrouter.jisc.ac.uk/api/v4/notification?api_key=${process.env.PUBROUTER_API_KEY}`;
+
+  const response = await fetch(apiEndpoint, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify(pdfMetadata),
+  });
+
+  return response.json();
+};
+
+const retryPostingPublication = async (pdfMetadata) => {
+  await new Promise((resolve) => setTimeout(resolve, 5000));
+  return postToPubrouter(pdfMetadata);
+};
+
+const sendFailureEmail = async (error, event, publicationId) => {
+  const ses = new AWS.SES();
+
+  const emailParams = {
+    Destination: {
+      ToAddresses: [process.env.EMAIL_RECIPIENT],
+    },
+    Message: {
+      Body: {
+        Text: {
+          Data: `Publication Id: ${publicationId}, Error: ${error}, Event: ${event}`,
+        },
+      },
+      Subject: {
+        Data: "A PubRouter delivery has failed.",
+      },
+    },
+    Source: "noreply@octopus.com",
+  };
+
+  return ses.sendEmail(emailParams).promise();
+};
+
+const mapPublicationToMetadata = (publication, pdfUrl) => {
+  const formatAuthor = (author, coAuthor) => {
+    return {
+      type: coAuthor ? "author" : "corresp",
+      name: {
+        firstname: author.firstName,
+        surname: author.lastName,
+        fullname: `${author.lastName}, ${author.firstName}`,
```

Comment:
Possibly worth checking if lastName is set here before using the ", " since although most real people will have one it is optional in our User model.

Reply:
Good spot, thanks!